### PR TITLE
The rest of std::string's API

### PR DESCRIPTION
**Pull request by SSoulaimane**
Implemented:
* atomic ref counting in the GCC version.
* growing with `resize()`.
* `reserve()`, `shrink_to_fit()`, `insert()`, `replace()`, and `swap()`.

### CODE REVIEW

**Comment by SSoulaimane** [comment]
Tested manually on linux, windows and osx (with https://github.com/dlang/dmd/pull/10593), and the tests are passing. On linux, only the old ABI is working which can be turned on with `version = _GLIBCXX_USE_CXX98_ABI`, the c++11 ABI is practically unusable until there is a move constructor in D, and issue [20321](https://issues.dlang.org/show_bug.cgi?id=20321) is closed.

File: src/core/internal/lifetime.d [hunk 1]
```diff
@@ -107,3 +107,15 @@ void emplaceInitializer(T)(scope ref T chunk)
...
+/*
+Simple swap function.
+*/
+void swap(T)(ref T lhs, ref T rhs)
+{
+    import core.lifetime : move;
+
+    T tmp = move(lhs);
+    lhs = move(rhs);
+    rhs = move(tmp);
```
Comment (kinke) [inline src/core/internal/lifetime.d]: This should be slightly more efficient:
```D
T tmp = move(lhs);
moveEmplace(rhs, lhs);
moveEmplace(tmp, rhs);
```

Comment (SSoulaimane) [reply src/core/internal/lifetime.d]: Definitely.

**Comment by thewilsonator** [comment]
This is green. Is it good to go?

**Comment by SSoulaimane** [comment]
It should be good to go unless a mistake is discovered. Note that OSX tests are still disabled and that dlang/dmd#10593 is still a prerequisite.

**Pull request merged**